add task: Developing and debugging services with local IDE inside container (#28564)
kubernetes/website: content/en/docs/tasks/debug-application-cluster/inside-container-developing.md

---

> title: Developing and debugging services with local IDE inside container

Reviewer: 这个表达很有问题。
是 IDE 在容器中么？

Reviewer: 后文的英文表达问题太多，请寻求英文较好的同事帮助重写。

Author: 好的，我重写一下

---

> title: Develop and debug the services in the container

Reviewer 2: ```suggestion
title: Developing and debugging applications running in your cluster
```

content_type: task
---

Kubernetes applications are generally composed of many microservices. Due to environment differences and resource constraints, these microservices somtimes can not run locally as source code. Therefore, developers need to rebuild the image, modify the image version of the workload and then wait for the Pod scheduling after coding. This causes the low efficiency of development in Kubernetes emvironment.

> 

Reviewer 2: ```suggestion

This page describes how you can external tools to debug an application that's running in your cluster.
The tools covered are:
{{% thirdparty-content %}}

- [Nocalhost](https://nocalhost.dev)

```

It would be helpful to make a similar change to the Telepresence task, so that they match each other.

Nocalhost is a tool that enables developers to develop applications in Kubernetes Cluster. It can transfer the remote workload to development mode by using Nocalhost IDE (including VSCode and Jetbrains). When the developer writes code locally, any code changes will be automatically synchronized to the remote development container and initiate the application update immediately (depending on the hot load mechanism of the application or restart the application). The development container inherits all the declarative configurations (configmap, secret, volume, Env) from the original workload, which provides the basis for the application to run in the development container as source code.

This is an tutorial of developing services in the container in Kubernetes environment using  `Nocalhost` .

## {{% heading "prerequisites" %}}

> * Kubernetes cluster is installed

Reviewer 2: ```suggestion
* {{< include "task-tutorial-prereqs.md" >}}
```

> Nocalhost is a tool that enables developers to develop applications in Kubernetes Cluster. It can transfer the remote workload to development mode by using Nocalhost IDE (including VSCode and Jetbrains). When the developer writes code locally, any code changes will be automatically synchronized to the remote development container and initiate the application update immediately (depending on the hot load mechanism of the application or restart the application). The development container inherits all the declarative configurations (configmap, secret, volume, Env) from the original workload, which provides the basis for the application to run in the development container as source code.

Reviewer: I still believe this post needs some work. For example, the paragraph above could be revised:

```suggestion
Nocalhost is a tool that enables developers to write applications to run on a Kubernetes Cluster.
By using a Nocalhost IDE (for example, VSCode or Jetbrains), you can switch the remote workload
into development mode. When code is updated on your local machine, changes to code are automatically
synchronized to the remote development environment. The code change then automatically triggers
an update to the target application. Depending on the nature of the application, such an update could
be a hot reload or a process restart. The development container where the updated application runs
inherits all the configurations (e.g. ConfigMap, Secret, Volume, Env) from the original container.
```

I'm not sure what a "development container" means and I'm curious how the target workload/application
is automatically restarted.

Reviewer: I'm not a native speaker, so the rewriting above only serves an example how the text can be reorganized. It is only an example to show how heavily this post should be revised. Without such a rewriting, this post is difficult to understand.

Author: Thank you! 
I have explained about the "development container" and rewritten it according to your comments.

> *  [Nocalhost](https://nocalhost.dev/installation/) extension installed (supporting VSCode and Jetbrains)

Reviewer 2: ```suggestion
* A local IDE: either Microsoft Visual Studio Code, or Jetbrains
* Nocalhost itself. To learn how to install Nocalhost, visit [Installation](https://nocalhost.dev/eng/installation/)
```

## Add Kubeconfig for Nocalhost 

> We take the VSCode extension as an example here. Open the Nocalhost extension, then you can add the Kubernetes Cluster by selecting the context in `~/.kube/config` or copying the "kubeconfig".  

Reviewer 2: I wouldn't be confident I could explain these steps to someone else. What should a reader do?

## Install the Demo application 

After adding the cluster, open it and it will displays all  `Namespace`. Find the  `default` Namespace and click the right button ![](/images/docs/nocalhost-deploy.png).

Select "Deploy Demo" to install the "bookinfo" application of lstio examples.

> ![Install bookinfo demo](/images/docs/nocalhost-demo.png)

Reviewer 2: This sounds like a task prequisite (you don't need it to perform debugging, but the task doesn't make sense without it). I'd explain this earlier.

After the application is installed, open the "bookinfo" application and enter the "Deplpoyment" menu. When all service icons turn green, the application becomes ready.

> ![Wait bookinfo becomes ready](/images/docs/nocalhost-bookinfo.png)

Reviewer 2: “Wait bookinfo becomes ready” isn't correct English; can you rewrite it?

Next, open `127.0.0.1:39080` in the browser to access the `productpage` service. You will see the following interface.

![Bookinfo productpage](/images/docs/nocalhost-bookinfo-productpage.png)

> ## Develop service

Reviewer 2: ```suggestion
## Modifiying your application
```

* The first method is to modify the source code, and then execute `docker build` to build the image. Push it to the image registry, update the image version of the workload and wait for it to work.
* The second method is to use tools such as  `Telepresence`. Run `productpage` service locally as source code and intercept the local domain name resolution to the remote cluster to access the  `details`, `ratings`, `reviews` services of the remote.

For some complex microservices, it is very difficult to run locally as source code due to environment differences and resource constraints. Hence, we have no choice but go back to the first method, which is the current situation of many complex application development.

But now, with Nocalhost, you have a new choice: develop the service in the container.

It becomes very easy to develop `productpage` services. Click the ![](/images/docs/nocalhost-start-develop.png) icon on the right side of the service on the extension, and select "Clone from Git Repo" in the pop-up menu and then choose a local directory. Nocalhost will atomatically clone the source code of this service and enter its development mode.

> In development mode, VSCode will open the source code in a new window and obtain access to the remote container  `Terminal` .

Reviewer 2: Is VS Code a strict prerequisite? It's not clear?

![Nocalhost remote dev container](/images/docs/nocalhost-remote-dev-container.png)

Execute `ls` command in the  `Terminal` :

```
/home/nocalhost-dev# ls
Dockerfile  README.md  productpage.py  requirements.txt  run.sh  static  templates  test-requirements.txt  tests

> ```

Reviewer 2: [Don't include the prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt)

Author: done

You can find that the source code of  `productpage` has been synchronized to the container. Then execute the command:

```
sh run.sh
```

The `productpage` service will start in a moment .

Next, use `VSCode` to edit the `355` line of  `productpage.py` file in the current workplace. Change `The Comedy of Errors` to `The Comedy of Errors Code Change Here`, and save the changes.

![Productpage Code](/images/docs/nocalhost-bookinfo-productpage-code-change.png)

> Refresh the pages of `127.0.0.1:39080`. You will see the following interface.

Reviewer 2: ```suggestion
In your web browser, refresh the window or tab where you have `http://127.0.0.1:39080/` open. You will see the following interface:
```

![Productpage after code changed](/images/docs/nocalhost-bookinfo-productpage-new-web.png)

Now, any code changes will be automatically synchronized to the remote. Since the `productpage` service has the hot reload ablity, the changes will work with immediate effect.

So far, we have finished the demostration of the `productpage` service development using Nocalhost.

> There is no need for developers to learn extra commands and tools in this process. It provides a visualized way to develop services in the container through developer's IDE, which greatly improves the experience and efficiency of development in Kubernetes environment.

Reviewer 2: I'd omit this text. It's not helping the reader learn a task.

## {{% heading "whatsnext" %}}

> If you are interested in this tutorial, please visit [quick start](https://nocalhost.dev/getting-started/).
> 
> For more information, please visit  [Nocalhost Website](https://nocalhost.dev).

Reviewer 2: From the Kubernetes project's point of view, we'd want to signpost readers to:
```suggestion
* Learn about [Developing and debugging services locally](/docs/tasks/debug-application-cluster/local-debugging/)
```
